### ISSUE TITLE
First-class support for conditional resources

### ISSUE THREAD

**Issue by jan-g** [hashicorp/terraform#25406]
TF's expressivity has seen a massive improvement since 0.11 -> 0.12. This is a small proposal for a minor improvement that reduces the need for state-file edits with some tf refactors.

### Current Terraform Version

All (this is a feature request).

### Use-cases

Currently, the conditional declaration of a resource in TF requires a specific idiom:

```
resource foo bar {
  count = condition ? 1 : 0
  etc
}
```

The issue with this is that the index occurs in the terraform address if the resource: the above would be named `foo.bar[0]`.

This can cause unnecessary churn and a need for state-file editing if a tf refactor needs to make a resource conditional (or remove that conditionality).

### Attempted Solutions

As above. There's an idiom for this but it's ugly and requires state-file edits to support a code refactor that adds or removes conditionality from a resource.

### Proposal

It would be great to see something that provides first-class support for this - perhaps along the lines of

```
resource foo bar {
  if = condition
}
```

where the resulting resource was merely named `foo.bar` (or didn't exist in the state file).

(We'd have a use for this both on individual resources *and* modules.)

### References

(None I can see.)

**Comment by apparentlymart** [comment]
Hi @jan-g! Thanks for this proposal.

I think the main thing that's come up when this has been discussed in the past is what exactly `foo.bar` ought to return when the condition is false in `resource "foo" "bar`".

Other Terraform language features are designed to work well with lists that might be empty, so conditionally setting `count` to either zero or one composes well with those features. A resource with a singular condition set to false could potentially appear as `null` when referenced, but the rest of the language doesn't have as many features for expressively dealing with values that might be null and so it's likely then that the rest of the configuration would end up worse off, replacing expressions like this:

```hcl
foo.bar[*].id
```

...with expressions like this:

```hcl
foo.bar != null ? foo.bar.id : null
```

This seems it also means that adding a conditional would still require changes elsewhere in the module: one way or another, the rest of the module must contend with the fact that something that was previously always set is now conditionally set, and so I expect it'd be rare that you could _just_ add the `if` argument to a resource and change nothing else in the module.

Aside from that specific practical concern, there is also a complexity tradeoff to be made here: is the value of this new feature high enough to justify the the cost of having yet another reserved argument inside resource blocks for users to understand, and having to establish another parallel set of usage patterns for referring to conditional resources? So far we have concluded that the answer is no: the need to deal with the conditional object as a list has felt like a small price to pay, because the Terraform language features tend to make it succinct to deal with in common cases.

The addition of `count` and `for_each` for modules in Terraform 0.13.0 is likely to affect this equation too, since it's likely to make it far less common for modules to take conditional flags that must then be handled on every resource in the module and more common to write the conditional only once on the module itself. The "overhead" of dealing with a list of zero or one items is therefore limited only to accessing the module outputs externally, and not to every resource in the module separately.

With all of that said, I think we'll at least need to wait to see how the module `count` and `for_each` features change the context around this decision. We'll see how this plays out in the coming months as folks start to make use of the module repetition features.

Thanks again for making this proposal!

**Comment by michelzanini** [comment]
Hi @apparentlymart,

Regarding protecting for nulls with:

```
foo.bar != null ? foo.bar.id : null
```

You could then also add a feature present in languages like Groovy using the operator `?`:

```
foo.bar?.id
```

This is a "null safe walk". It means if `bar` is `null` then the whole expression becomes `null`.

I know its another thing to add to the language but it would make just much more elegant and eliminate those resource[0] on state files.

**Comment by jan-g** [comment]
FWIW, our uses involve conditional "branches" of configuration - which is to say, the things we're turning on or off look like "forks" in the dependency tree, sans any "join". The thing I'm looking to support here is the use of the same TF to stamp out a dev environment (which may have a small amount of additional or alternative machinery) as well as staging and prod.

We'd probably be applying these conditionals at the module level, as @apparentlymart suggests. It still means that refactorings that add or remove conditionals (counts of 0 or 1) require a `terraform state mv` dance, which is the thing I'm hoping to avoid.

Our use case could survive a constraint such as "no sibling-level references to conditional resources" but I suppose it's possible that there are those which cannot: assembling IP addresses from a set of hosts, some of which may be conditionalised, to hand-wave at the kind of thing where this could arise. I'd think that you would indeed require some syntax that makes a reference to a conditional resource behave like "a bag of 0 or 1 things" - that feels more in keeping with TF, and has a consistent type profile, than adding optional values.

**Comment by kungfuchicken** [comment]
Our use case is similar to the one described by @jan-g. The value I'm lacking and looking for in a feature like this is improving the ability to refactor code around conditional resources. I could see this implemented in a language-level meta-argument, or as an improvement in the tools that execute the code. If a `terraform plan` was smart enough to realize that a sudden add/remove of a count argument is a developer using the `count = boolean` pattern and do the required state modification without requiring manual intervention, then that could also deliver value. There could possibly even be flags on the command to facilitate this, or some sort of syntactic sugar that provides it at a language level, but implements it equivalent to a `count = boolean`. 

The `count = boolean` is a clever hack. It's not a true conditional, Like many things, there are tradeoffs to different approaches, but it seems like there is potential value being missed out on.

**Comment by kungfuchicken** [comment]
Also, the argument "but if you have conditionals you'll need to touch other parts of the code" kind of glosses over that the same is true if there are existing references to resources/modules with/without a count.  Since the structure of the state has changed, the references need to be updated as well. Potato potato ¯\_(ツ)_/¯.

**Comment by jcmcken** [comment]
We frequently run into this issue as well. The most common case is where we have a module that always creates a specific resource (let's call it resource A) and then we later want to expand the module to make it optional via a boolean variable that defaults to true (so that module consumers can opt not to create resource A). The behavior of the updated module, if the consumer does not do anything else, should be the same as the previous behavior. Naively, it seems like a very simple code change. But because the shape of the state changes, it turns out to be incredibly tedious.

In our environment, we have thousands and thousands of resource declarations. Going through all of the prior consumers of the module and fixing their state takes forever. Deleting and recreating is often not possible (or at least is prohibitive) because the resource is stateful in nature (e.g. an AWS RDS database that has lots of data -- migrating terabytes of data to new RDS instances just because the state's shape is changing is just not feasible).

Perhaps Terraform could just "alias" a singleton resource so that it can be accessed as an array or as a singleton. For example, if there's a singleton resource ``foo.bar``, but I access it as ``foo.bar[0]`` (or vice versa) Terraform understands what I mean and does not error (and if I access it as ``foo.bar[1]``, Terraform does error because a singleton can't be represented by an array of length 2). In the case of comparing current state to stored state, Terraform could do a similar translation, and if the access is successful, a state change is deemed unnecessary. Downside to this approach is the typing system is getting muddied a bit.

Or, in a similar vein, Terraform could just always normalize resources as belonging to an array. So if I access with the singleton name, it translates it to the array name and that's how it's stored in the state. This option seems more distasteful, though, since it necessitates updates to the state, which is what people in this ticket are trying to avoid. I suppose some kind of migration tool could help workaround that issue.

Or just support native conditionals instead of requiring people to use "count"

**Comment by apparentlymart** [comment]
Hi all,

I see some mentions here of having to use `terraform state mv` after adding or removing a `count` argument from a resource. I'm surprised by that _particular_ observation, because going back to before I was even working on Terraform there has been a special rule that Terraform will implicitly and automatically rename a single-instance to a zeroth instance and vice-versa when adding or removing `count`. If you're seeing that not be true, please open an new issue for that because that's a bug and something separate from this proposal.

I want to note that the same doesn't apply to `count` when added or removed from a `module` block. Module repetition is a far more complex beast that doesn't really lend itself well to automatic adjustments like that, so you do still need to handle that change explicitly. However, those with that concern might want to have a look at [the config-driven refactoring experiment](https://discuss.hashicorp.com/t/request-for-feedback-config-driven-refactoring/30730/3) for which we're currently asking for feedback. If the feedback on this looks good we're hoping to release it as stable in a v1.1 release in the not-too-far future, at which point it'll be possible to write the equivalent of a `terraform state mv` call directly inside your module and have Terraform automatically apply it as part of the next plan.

(The automatic adjustment for resources I mentioned above is also incorporated into this new feature, where now Terraform will treat it as a sort of implicit `moved` block, rather than as an unlike-anything-else special case as before.)

**Comment by apparentlymart** [comment]
Hi again all!

Lately I've noticed that this request is captured in a few different issues, which is causing the design discussion to get fragmented. I'm going to close this one in favor of #21953 because that one has the most existing discussion and the most existing :+1: upvotes.

If you currently have an upvote attached to this issue then I suggest transferring it over to #21953 so we can track all of that in a central place. Thanks!
